### PR TITLE
Fixing #1402

### PR DESCRIPTION
**Pull request by KKhurin**
Fixing #1402
Removing IssueAttribute from failing session tests
Exposing OperationContext.GetCallbackChannel in the contract

### CODE REVIEW

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ServiceChannel.cs [hunk 1]
```diff
@@ -995,6 +1004,14 @@ private void HandleReply(ProxyOperationRuntime operation, ref ProxyRpc rpc)
                                             remoteAddress);
             }
         }
+        void TerminateIfNecessary(ref ProxyRpc rpc)
```
Comment (shmao) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ServiceChannel.cs]: Specify the visibility.
Nit: add an empty line.

File: src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/terminatingoperationbehavior.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (shmao) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/terminatingoperationbehavior.cs]: Nit: `CamelCase` the file name.

File: src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/TerminatingOperationBehavior.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+            return false;
+        }
+
+        internal void AfterReply(ref MessageRpc rpc)
+        {
+            if (rpc.Operation.IsTerminating && rpc.Channel.HasSession)
+            {
+                var timer = new Timer(TerminatingOperationBehavior.AbortChannel, rpc.Channel.Binder.Channel, rpc.Channel.CloseTimeout, TimeSpan.FromMilliseconds(-1));
+            }
```
Comment (mconnew) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/TerminatingOperationBehavior.cs]: Wrap AbortChannel in UnhandledExceptionFrame wrapper

File: src/System.ServiceModel.Primitives/tests/Description/OperationContractAttributeTest.cs [hunk 4]
```diff
@@ -34,6 +34,18 @@ public static void IsInitiating_Method_Requires_Proper_SessionMode()
...
+        // We no longer ignore IsTerminating = true in our contract invariant checks 
+        // so they will now fail if the ServiceContract's SessionMode isn't SessionMode.Required
+        var exception = Assert.Throws<InvalidOperationException>(() =>
+        {
+            channelFactory.CreateChannel();
+        });
+        Assert.True(exception.Message.Contains("IInitiatingTerminatingService"));
+    }
     static ChannelFactory<T> CreateChannelFactoryHelper<T>(string url)
     {
```
Comment (mconnew) [inline src/System.ServiceModel.Primitives/tests/Description/OperationContractAttributeTest.cs]: Nit: New line after method close

File: src/System.ServiceModel.Primitives/ref/System.ServiceModel.Primitives.cs [hunk 5]
```diff
@@ -528,6 +528,7 @@ public sealed partial class OperationContext : System.ServiceModel.IExtensibleOb
         public System.ServiceModel.Channels.MessageProperties OutgoingMessageProperties { get { return default(System.ServiceModel.Channels.MessageProperties); } }
         public System.ServiceModel.Channels.RequestContext RequestContext { get { return default(System.ServiceModel.Channels.RequestContext); } set { } }
         public event System.EventHandler OperationCompleted { add { } remove { } }
+        public T GetCallbackChannel<T>() { return default(T); }
     }
```
Comment (mconnew) [inline src/System.ServiceModel.Primitives/ref/System.ServiceModel.Primitives.cs]: Please create issue for needing a test for this.

File: src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/TerminatingOperationBehavior.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        private static void AbortChannel(object state)
+        {
+            try
+            {
+                ((IChannel)state).Abort();
+            }
+            catch (SocketException)
+            {
+                // this is a timer callback so we protect it against SocketException from the call to Abort()
+            }
```
Comment (mconnew) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/TerminatingOperationBehavior.cs]: SocketException wouldn't be thrown from calling Abort on all types of channels.

File: src/System.ServiceModel.Primitives/ref/System.ServiceModel.Primitives.cs (hunk 5 above)
Comment (KKhurin) [inline src/System.ServiceModel.Primitives/ref/System.ServiceModel.Primitives.cs]: #1902 - this will likely get covered by duplex session tests.

Comment (KKhurin) [inline src/System.ServiceModel.Primitives/ref/System.ServiceModel.Primitives.cs]: Also added Channel property to the contract

File: src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/TerminatingOperationBehavior.cs (hunk 3 above)
Comment (KKhurin) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/TerminatingOperationBehavior.cs]: Yep.

File: src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/TerminatingOperationBehavior.cs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+            ((IChannel)state).Abort();
+        }
+
+        public static TerminatingOperationBehavior CreateIfNecessary(DispatchRuntime dispatch)
+        {
+            if (IsTerminatingOperationBehaviorNeeded(dispatch))
+            {
+                return new TerminatingOperationBehavior();
+            }
+            else
```
Comment (mconnew) [inline src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/TerminatingOperationBehavior.cs]: Redundant else clause.

**Pull request merged**